library/std/src/sys/exit_guard.rs

```diff
@@ -31,27 +25,29 @@ cfg_if::cfg_if! {
         #[cfg_attr(any(test, doctest), allow(dead_code))]
         pub(crate) fn unique_thread_exit() {
             let this_thread_id = unsafe { libc::pthread_self() };
-            use crate::sync::{Mutex, PoisonError};
-            static EXITING_THREAD_ID: Mutex<Option<PThread>> = Mutex::new(None);
-            let mut exiting_thread_id =
-                EXITING_THREAD_ID.lock().unwrap_or_else(PoisonError::into_inner);
-            match *exiting_thread_id {
-                None => {
+
+            const _: () = assert!(mem::size_of::<libc::pthread_t>() <= mem::size_of::<usize>());
+
+            const NONE: usize = 0;
```

Comment:
Have you confirmed that zero is definitely not an allowed `pthread_t` value?

Reply:
According to https://man7.org/linux/man-pages/man3/pthread_equal.3.html:

> POSIX.1 allows an implementation wide freedom in choosing the
       type used to represent a thread ID; for example, representation
       using either an arithmetic type or a structure is permitted.
       Therefore, variables of type pthread_t can't portably be compared
       using the C equality operator (==); use [pthread_equal(3)](https://man7.org/linux/man-pages/man3/pthread_equal.3.html) instead.

I don't think it's a good idea to assume `pthread_t` can be trivially compared as integers.

Reply:
Yeah, in retrospect I should have insisted on that in #126606. We only use this on Linux, so this currently works. In the long-term, I think `thread::current().id()` is the best solution, but that's currently not available in TLS destructors (#124881 will improve that situation and I have an idea for solving this completely).

Reply:
> I don't think it's a good idea to assume `pthread_t` can be trivially compared as integers. I can't immediately tell if the original implementation uses integer equality, but probably [`pthread_equal`](https://man7.org/linux/man-pages/man3/pthread_equal.3.html) (or a PartialEq implementation based on that) should be used instead.

AFAIK the passage you quoted was added because C can't use `==` for `struct`s. We can see that none of the targets supported by `libc` use a `struct` for `pthread_t`. I think we should disregard this section of the man page.

Reply:
I used `==` in [the original PR](https://github.com/rust-lang/rust/pull/126606#discussion_r1664617977) mostly just because `libc` does not (yet) expose `pthread_equal`. glibc and musl at least both appear to implement `pthread_equal` as just `==` from what I found.

Reply:
> thread::current().id()

[iirc](https://github.com/rust-lang/rust/issues/67939#issuecomment-2120475942) `thread::current()` does heap alloc(due to internal use of `Arc`) which would be bad if this process is a forked process, or wouldn't it?

<details>

> "After  a  fork()  in a multithreaded program, the child can safely call only async-signal-safe functions (see signal-safety(7)) until such time as it calls execve(2)." - from `man 2 fork`

`malloc` isn't in the list of functions that are async-signal-safe

</details>

Reply:
`exit` isn't signal-safe, so it mustn't be called after `fork` anyway.

Reply:
I updated #127912 to include a change that makes `thread::current_id` always succeed and always return the same ID. Once that gets merged, we can use it here.